desc/objectives/_coils.py

```diff
@@ -0,0 +1,591 @@
+        Weighting to apply to the Objective, relative to other Objectives.
+        Must be broadcastable to to Objective.dim_f
+    normalize : bool, optional
+        Whether to compute the error in physical units or non-dimensionalize.
+    normalize_target : bool, optional
+        Whether target and bounds should be normalized before comparing to computed
+        values. If `normalize` is `True` and the target is in physical units,
+        this should also be set to True.
+        be set to True.
+    loss_function : {None, 'mean', 'min', 'max'}, optional
+        Loss function to apply to the objective values once computed. This loss function
+        is called on the raw compute value, before any shifting, scaling, or
+        normalization. Note: Has no effect for this objective.
+    deriv_mode : {"auto", "fwd", "rev"}
+        Specify how to compute jacobian matrix, either forward mode or reverse mode AD.
+        "auto" selects forward or reverse mode based on the size of the input and output
+        of the objective. Has no effect on self.grad or self.hess which always use
+        reverse mode and forward over reverse mode respectively.
+    grid : Grid, list, optional
+        Collocation grid containing the nodes to evaluate at. If list, has to adhere to
+        Objective.dim_f
+    name : str, optional
+        Name of the objective function.
+
+    """
+
+    def __init__(
+        self,
+        coil,
+        data_keys,
+        target=None,
+        bounds=None,
+        weight=1,
+        normalize=True,
+        normalize_target=True,
+        loss_function=None,
+        deriv_mode="auto",
+        grid=None,
+        name=None,
+    ):
+        self._grid = grid
+        self._data_keys = data_keys
+        self._normalize = normalize
+        super().__init__(
+            things=[coil],
+            target=target,
+            bounds=bounds,
+            weight=weight,
+            normalize=normalize,
+            normalize_target=normalize_target,
+            loss_function=loss_function,
+            deriv_mode=deriv_mode,
+            name=name,
+        )
+
+    def build(self, use_jit=True, verbose=1):  # noqa:C901
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        # local import to avoid circular import
+        from desc.coils import CoilSet, MixedCoilSet, _Coil
+
+        self._dim_f = 0
+
+        def get_dim_f(coilset):
+            """Get dim_f."""
+            if isinstance(coilset, list):
+                [get_dim_f(x) for x in coilset]
+            elif isinstance(coilset, MixedCoilSet):
+                [get_dim_f(x) for x in coilset]
+            elif isinstance(coilset, CoilSet):
+                get_dim_f(coilset.coils)
+            elif isinstance(coilset, LinearGrid):
```

Comment:
does this ever get called on a coilset? seems like it would just return None? Since the bottom of the recursion is a `Coil` object, not a `LinearGrid`

Reply:
No this only gets called on a `MixedCoilSet` or `CoilSet` container containing LinearGrids (`self._grid`). Since we're using `tree_map` to map the structure of `self.things[0]` to the inputted grid, the inputted grid is contained within a `CoilSet` or `MixedCoilSet`. e.g. if I input `CoilSet(coil, n=2)`, then `coilset.coils` is `[LinearGrid, LinearGrid]`